include/pybind11/numpy.h

```diff
@@ -639,11 +727,29 @@ class dtype : public object {
     /// Single character for byteorder
     char byteorder() const { return detail::array_descriptor_proxy(m_ptr)->byteorder; }
 
-    /// Alignment of the data type
+/// Alignment of the data type
+#ifdef PYBIND11_NUMPY_1_ONLY
     int alignment() const { return detail::array_descriptor_proxy(m_ptr)->alignment; }
+#else
+    ssize_t alignment() const {
+        if (detail::npy_api::get().PyArray_RUNTIME_VERSION_ < 0x12) {
+            return detail::array_descriptor1_proxy(m_ptr)->alignment;
+        }
+        return detail::array_descriptor2_proxy(m_ptr)->alignment;
+    }
+#endif
 
-    /// Flags for the array descriptor
+/// Flags for the array descriptor
+#ifdef PYBIND11_NUMPY_1_ONLY
     char flags() const { return detail::array_descriptor_proxy(m_ptr)->flags; }
+#else
+    std::uint64_t flags() const {
```

Comment:
Asking myself: What could possibly go wrong here? — I feel very uncertain.

If someone has code like `char flags = dtype.flags();`

Is it possible that the `flags` are truncated? Could that result in silent failures?

Does this idea have any merit?

```
char flags() const { ... return flags if only the lower 8 bits are non-zero, or throw exception otherwise ... }
std::uint64_t flags64() const { ... always succeeds / never throws ... };
```

Reply:
True, but nobody in current code can possibly use a flag that would be lost (they don't exist currently).
Additionally, you would have to write `char flags = `, rather than `(dtype.flags() & FLAG) == FLAG` or so to begin with, which seems unlikely.

So yes, it's plausible, but basically a NumPy problem not a pybind11 one, and to me it seems unproblematic enough.

Reply:
That sounds fine to me. A compiler should warn about truncation if that’s happening, as well.

Reply:
> A compiler should warn about truncation if that’s happening, as well.

It may or may not, and as someone working with a large codebase: seeing thousand of warnings nobody cared about enough is the norm. This will be just another needle in the haystack.

> So yes, it's plausible,

I see, thanks.

> basically a NumPy problem not a pybind11 one

Got it. Small ouch, but I agree there isn't anything we can do here. It seems to be "too late" already.